src/internal_rep.rs

```diff
@@ -2578,6 +2585,23 @@ fn validate_argument<'a>(
                 )));
             }
 
+            if let Some(resource_type) = types.get(constants::RESOURCE) {
+                if (target_argument.param_type == resource_type
+                    || target_argument
+                        .param_type
+                        .is_child_or_actual_type(resource_type, types))
+                    && arg_typeinfo.name == constants::SELF
+                    && !is_avc
+                {
+                    return Err(ErrorItem::Compile(CompileError::new(
+                        "'self' passed as resource argument",
+                        file,
+                        arg.get_range(),
+                        "'self' cannot be passed into a function as a resource since 'self' is dependent on local context",
```

Comment:
Thanks, this is nice.

Looking at the test case, I could imagine someone making this mistake because they wanted to use "this" instead.  What do you think about adding a parenthetical note that says "(If you intended to refer to the type declared in the enclosing block, use "this")"?